### PR TITLE
Correct Sys.os_type on 4.08.0 (Closes #122)

### LINKED ISSUE #122: Cross-compiler has `ocamlc -config-var os_type` = `"%%OCAML_OS_TYPE%%"`
I am trying to set up a toolchain with MXE (basically for GTK+3), and `x86_64-w64-mingw32.static`.

I can install the `ocaml-windows64` fine, but then, when trying to compile my project with dune (`dune build @install -x windows`, some linking options are clearly wrong (`-I /usr/include/...`), and the compilation fails.

I am not sure if this is the cause, but I noticed that
```
$ ./_opam/bin/ocamlc -config-var os_type
Unix
$ ./_opam/windows-sysroot/bin/ocamlc -config-var os_type
%%OCAML_OS_TYPE%%
$ wine64 ./_opam/windows-sysroot/bin/ocamlc.exe -config-var os_type
%%OCAML_OS_TYPE%%
```
which causes `dune` to choose the wrong alternative when some files depend on `os_type = Win32`.

It seemed to be a bug in the `ocaml-windows64` package, so I attempted to fix it (https://github.com/AltGr/opam-cross-windows/commit/99b9649283e95487091117628bff5a1174c2c8f0).

However, with this fix OCaml no longer compiles, with
```
.../_opam/bin/ocamlrun ../boot/ocamlc -use-prims ../runtime/primitives -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats   -o stdlib__pervasives.cmo -c pervasives.ml
File "..._opam/.opam-switch/build/ocaml-windows64.4.08.0/stdlib/pervasives.ml", line 19, characters 18-29:
19 | let invalid_arg = invalid_arg
```
which is quite a surprising outcome. I am guessing that something got broken in the compiler (the compilation line is the same, and there are lots of `os_type = "Win32"` checks in ocaml's code) ; but I can't figure why, and I thought I would first ask if the above was expected ?

Any help welcome!

### PR DESCRIPTION
**Pull request by pirbo**

### CODE REVIEW

**Comment by toots** [comment]
Hi! Thanks for this contribution! Have you been able to test the patch? It seems very minimal to me. I've had issues with `Sys.os_type` reporting the wrong OS. Without testing, would you mind explaining how this makes it work?

**Comment by pirbo** [comment]
I'm sorry. I may have done things too fast.
My patch does not touch at Sys.os_type really. It simply ensure that the os_type returned by `ocamlc` is the os_type it compiles for instead of the os_type it runs on...

My test is: both `wine ~/.opam/4.08.0/windows-sysroot/bin/ocamlc.exe -config` and `~/.opam/4.08.0/windows-sysroot/bin/ocamlc -config` return the correct output:
```
[...]
os_type: Win32
[...]
```
Instead of `~/.opam/4.08.0/windows-sysroot/bin/ocamlc -config` returning without it:
```
[...]
os_type: Unix
[...]
```

(Works also in @AltGr case `~/.opam/4.08.0/windows-sysroot/bin/ocamlc -config-var os_type`

Our difficulties with os_type in this repo is the interaction of 2 things
- a smart but weird hack made to ensure that `Sys.os_type` is determined only at execution of the program. 
The use case being "program compiled in bytecode that uses the Unix module":
(The principle of bytecode compilation is that it is cross platform. So in Unix module, you need to be ready to run on any arch and to decide the implementation to use at runtime. That's why functions are written in the style
```
let something x y = match Sys.os_type with
| Unix -> unix_implementation_of_something
| Win32 -> win32_implementation_of_something
```
and that in Sys `os_type` comes from `get_config` something that comes from C and therefore the current `ocamlrun`you're using to run the program...)
- Our cross compiler is not really a cross compiler. It is a compiler for windows on windows launched through a native unix `ocamlrun`.

In conclusion,
1/ I mess up in the name of my PR, sorry.
2/ Clearly, the output of os_type done by `ocamlc -config` is expected to be the type of os the compiler compiles to. This is what my fix does by putting the value of os_type determined statically at compile time of the compiler instead of the value computed at runtime
3/ The question of what should Sys.os_type outputs in general is less clear to me...

**Comment by toots** [comment]
Thanks for this insightful and detailed comment. `Sys.os_type` for cross-compiling has been the number one thing I've been meaning to fix and commit to OCaml `trunk`. @shindere say he should be able to help pushing this through.

I'll have a full review of your comment and the code and, hopefully we can come up with a good solution.

**Comment by AltGr** [comment]
For the record, I have been continuing my tests using this patch and it does solve many issues. Thanks!

**Comment by toots** [comment]
Thanks for letting me know. I'll test it very shortly and will merge is it works for me as well.

**Comment by shindere** [comment]
Romain Beauxis (2019/10/02 08:24 -0700):
> Thanks for this insightful and detailed comment. `Sys.os_type` for
> cross-compiling has been the number one thing I've been meaning to fix
> and commit to OCaml `trunk`. @shindere say he should be able to help
> pushing this through.

Absolutely. If you submit such a PR please make sure to mention me
because I'm not watching the repository, to avoid receiving more e-mails
I am able to handle.

**Comment by toots** [comment]
@pirbo Do you know why https://github.com/ocaml-cross/opam-cross-windows/issues/169 is happening?  The full output is:
```
Sys.win32: false
Sys.unix: true
Sys.cygwin: false
Sys.os_type: Win32
```
I can confirm that both `wine ~/.opam/4.08.0+32bit/windows-sysroot/bin/ocamlc.exe -config` and `~/.opam/4.08.0+32bit/windows-sysroot/bin/ocamlc -config` return `os_type` as `Win32. 

it looks like the `{win32,unix,cygwin}` constant are being returned from a different part of the code. Also, interestingly, the issue does not happen with the 64 bit compiler.

**Comment by pirbo** [comment]
No, I don't. 
This is indeed quite mysterious as the implementation of Sys.win32 is `return Val_bool(0 == strcmp(OCAML_OS_TYPE,"Win32"));` while Sys.os_type is exactly `caml_copy_string(OCAML_OS_TYPE);` so I don't understand how we can end up in this situation! I'll try to dig further tomorrow...

**Comment by toots** [comment]
Thanks, that's much appreciated. I agree with you that this is really confusing. The only reasonable guess that I can have is that the constants are resolved via a different runtime symbol. 

I could hardcode these values but I think it would be nice to get to the bottom of it, it might surface some interesting things.
